libraries/SPI/src/SPI.cpp

```diff
@@ -78,73 +94,100 @@ void SPIClass::usingInterrupt(uint8_t interruptNumber)
 
 void SPIClass::beginTransaction(uint8_t _pin, SPISettings settings)
 {
-  if(_pin > SPI_CHANNELS_NUM)
+  uint8_t idx;
+
+  if(_pin > NUM_DIGITAL_PINS)
     return;
 
-  spiSettings[_pin].clk = settings.clk;
-  spiSettings[_pin].dMode = settings.dMode;
-  spiSettings[_pin].bOrder = settings.bOrder;
-  if(spiSettings[_pin].bOrder == MSBFIRST) {
-    spiSettings[_pin].msb = MSBFIRST;
+  idx = getIdx(_pin);
+  if(idx == NB_SPI_SETTINGS) {
+    idx = addPin(_pin);
+    if(idx == NB_SPI_SETTINGS)
+      return;
+  }
+
+  spiSettings[idx].param.clk = settings.clk;
+  spiSettings[idx].param.dMode = settings.dMode;
+  spiSettings[idx].param.bOrder = settings.bOrder;
+  if(spiSettings[idx].param.bOrder == MSBFIRST) {
+    spiSettings[idx].param.msb = MSBFIRST;
   } else {
-    spiSettings[_pin].msb = LSBFIRST;
+    spiSettings[idx].param.msb = LSBFIRST;
   }
 
-  spi_init(&_spi, spiSettings[_pin].clk, spiSettings[_pin].dMode, spiSettings[_pin].msb);
-  g_active_id = _pin;
-}
+  if((_pin != CS_PIN_CONTROLLED_BY_USER) && (_spi.pin_ssel == NC)) {
+    pinMode(_pin, OUTPUT);
+    digitalWrite(_pin, HIGH);
+  }
 
-void SPIClass::endTransaction(void)
-{
-  g_active_id = -1;
+  spi_init(&_spi, spiSettings[idx].param.clk,
+                  spiSettings[idx].param.dMode,
+                  spiSettings[idx].param.msb);
+  _CSpin = _pin;
 }
 
-void SPIClass::end(uint8_t _pin)
+void SPIClass::endTransaction(uint8_t _pin)
 {
-  UNUSED(_pin);
-  end();
+  if(_pin > NUM_DIGITAL_PINS)
+    return;
```

Comment:
same as above

Reply:
This is also an error because _pin must be a defined pin (presents in digitalPin[]).